### PR TITLE
Modification stack to nodes

### LINKED ISSUE #4863: Remove SkeletonModificationStack and convert IK solvers to nodes.
### Describe the project you are working on

RenIK, a skeleton full body IK implementation.

### Describe the problem or limitation you are having in your project

SkeletonModification3D and SkeletonModificationStack3D are doubly-nested (two layers deep) which makes it hard to work with. And analogously for Skeleton2D

Further, because they are resources and were holding runtime state, it led to unusual glitches when duplicating skeletons and in some cases required restarting the editor when testing.

The fact is that the current implementation has bugs, and *some* change is needed. The open question is what type of change.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Skeleton Modifications will be replaced with Nodes, and the SkeletonModificationStack resources will be removed as well as the property in Skeleton3D / Skeleton2D.

Additionally, IK solvers will be changed to operate on the bone_pose instead of using overrides.

Nodes can be added anywhere in the tree, so they allow making use of the tree order for processing. For example, so some nodes could be added before an animation player and some after.

((Note: There is an alternative approach to nodes, which is to keep the Resources as they are, but use a RefCounted "Playback" object, like AnimationTree has. I think this would be more work, but it is the competing proposal.))

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Iteration order is not a problem because nodes update in tree order, so IK solvers can be composed simply by re-ordering children.

The `SkeletonModification3DStackHolder` could be implemented by making a child node with more children of its own.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

You could write IK implementations as nodes.

### Is there a reason why this should be core and not an add-on in the asset library?

Skeleton3D and Skeleton2D are core.

### PR DESCRIPTION
**Pull request by fire**
Fixes: https://github.com/godotengine/godot-proposals/issues/4863

Work in progress.

Using https://github.com/fire/godot-skeleton-modifier as the test project.

- [x] Change to set pose.
- [x] Use radians internally and degrees in the editor
- [ ] QA all the modifications by restoring the demo.
- [ ] Add a quality of life bindings of the bone suggestions.

---

List of issue possible closed by completing this implementation (SkeletonModificationStack issues related to 4.0):

- #41322
- #54234
- #54891
- #59554
- #60180
- #61259
- #62003
- #62876
- #63694
- #65167
- #65265
- #67669
- #69752

### CODE REVIEW

File: scene/3d/skeleton_modification_3d_fabrik.cpp [hunk 1]
```diff
@@ -619,10 +523,115 @@ void SkeletonModification3DFABRIK::_bind_methods() {
...
-	stack = nullptr;
 	is_setup = false;
 	enabled = true;
 }
 
 SkeletonModification3DFABRIK::~SkeletonModification3DFABRIK() {
 }
+
+void SkeletonModification3DFABRIK::_notification(int p_what) {
+	switch (p_what) {
```
Comment (fire) [inline scene/3d/skeleton_modification_3d_fabrik.cpp:534]: - set_process_internal / physics_process_internal -> superclass
- remove _cache functions everywhere
- get_node should be done in NOTIFICATION_READY
- take physics process and move it to the superclass
- set_execution_mode() should update the set_physics_process
- set_skeleton() should also get_node again
- delete is_setup and all related checks
- enabled -> set_process_internal(false) 

This is direct all like 10-15 modifications.

File: core/object/worker_thread_pool.cpp [hunk 2]
```diff
@@ -300,7 +300,7 @@ WorkerThreadPool::GroupID WorkerThreadPool::_add_group_task(const Callable &p_ca
 		group->tasks_used = 0;
 		p_tasks = 0;
 		if (p_template_userdata) {
-			memdelete(p_template_userdata);
+			p_template_userdata->~BaseTemplateUserdata();
```
Comment (fire) [inline core/object/worker_thread_pool.cpp:303]: Undo.

File: scene/3d/skeleton_modification_3d.h [hunk 3]
```diff
@@ -31,35 +31,26 @@
...
 #include "scene/3d/skeleton_3d.h"
-#include "scene/resources/skeleton_modification_stack_3d.h"
 
-class SkeletonModificationStack3D;
-
-class SkeletonModification3D : public Resource {
-	GDCLASS(SkeletonModification3D, Resource);
-	friend class Skeleton3D;
-	friend class SkeletonModificationStack3D;
+class SkeletonModification3D : public Node3D {
```
Comment (fire) [inline scene/3d/skeleton_modification_3d.h:37]: Should be type node. The use case of being a child of animation player.

File: scene/2d/skeleton_modification_2d.h [hunk 4]
```diff
@@ -32,38 +32,39 @@
...
 ///////////////////////////////////////
 
-class SkeletonModificationStack2D;
 class Bone2D;
 
-class SkeletonModification2D : public Resource {
-	GDCLASS(SkeletonModification2D, Resource);
-	friend class Skeleton2D;
-	friend class Bone2D;
+class SkeletonModification2D : public Node2D {
```
Comment (fire) [inline scene/2d/skeleton_modification_2d.h:42]: Should be type node. The use case of being a child of animation player.

**Comment by akien-mga** [comment]
CC @TwistedTwigleg

**Comment by fire** [comment]
@lyuma is working on the pr, so the current code isn't accurate. The proposal is still broadly accurate.

**Comment by TwistedTwigleg** [comment]
I don’t have the time to review the PR, but I agree with the general change (*not that my agreement is needed - it’s all Godot code now* 🙂 ). Moving it to nodes is MUCH closer to my original GSoC proposal and based on the work I did for my IK plugin for Godot 3.0, it is much more stable and easier to maintain as well. It should also make the workflow less confusing and reduce the number of clicks to get something working on screen.

Now just need to make those Bone3D nodes and have the modification nodes use the Bone3D nodes to fully fit the original proposal 😉 

In all seriousness though - I think the general change is good. It make the modification stack completely unnecessary and I think that is where a lot of the pain came from. It also will untie it from the Skeleton3D node code, which will make it standalone.

**Huge thanks to everyone who has continued to work on the IK stuff!** I haven’t had any time and energy at all to work on much in spare time as of late and so I’m glad and thankful that others have continued to work on it and improve it.

**Comment by fire** [comment]
The git branches for this work is at https://github.com/lyuma/godot/tree/modification-stack-to-nodes-wip-full and https://github.com/lyuma/godot/tree/modification-stack-to-nodes-wip

**Comment by lyuma** [comment]
@TwistedTwigleg I've updated this PR with a mostly feature-complete PR. This is now a two-commit PR. The reason is I don't know how to preserve the rename metadata if I make it all in one commit since so much code was rewritten, so it makes review easier to do it this way.

I haven't tested anything at all. There is likely a lot that could have broken, but the existing implementation already has many features broken during alpha.

As a future PR, there are still some methods that I would like to clean up in BoneAttachment3D related to update order and the pose_updated notification. I also find the external_skeleton feature to be confusing.

Regarding what is in this update... I added documentation for all the new and renamed methods/members.

@TokageItLab I addressed the biggest feedback items.
I removed the following functions:
> ```
>     void clear_bones_local_pose_override();
>     Transform3D get_bone_local_pose_override(int p_bone) const;
>     void set_bone_local_pose_override(int p_bone, const Transform3D &p_pose, real_t p_amount, bool p_persistent = false);
> 
>     void update_bone_rest_forward_vector(int p_bone, bool p_force_update = false);
>     void update_bone_rest_forward_axis(int p_bone, bool p_force_update = false);
>     Vector3 get_bone_axis_forward_vector(int p_bone);
>     int get_bone_axis_forward_enum(int p_bone);
> 
>     Basis global_pose_z_forward_to_bone_forward(int p_bone_idx, Basis p_basis);
> 
>     // Modifications
> #ifndef _3D_DISABLED
>     Ref<SkeletonModificationStack3D> get_modification_stack();
>     void set_modification_stack(Ref<SkeletonModificationStack3D> p_stack);
>     void execute_modifications(real_t p_delta, int p_execution_mode);
> #endif // _3D_DISABLED
> ```
get_bone_rest_forward_vector and global_pose_z_forward_to_bone_forward were migrated to SkeletonModification3D as helper functions.

I have made all calculations strictly based on `get_bone_pose` and `set_bone_pose`. Modification nodes do not depend on skeleton update or dirty and ignore global pose override.

I have removed bone_property_suffixes and used a `const_cast` to permit calling into GDScript from `_validate_property`. I would rather `_validate_property` be exposed the same way as `get_property_list` and others, but this should be good enough for now.

File: scene/2d/skeleton_modification_2d.cpp [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+		case NOTIFICATION_READY: {
+#ifdef TOOLS_ENABLED
+			Skeleton2D *skel = get_skeleton();
+			if (Engine::get_singleton()->is_editor_hint() && skel) {
+				skel->connect(SceneStringNames::get_singleton()->draw, Callable(this, SNAME("draw_editor_gizmo")));
+			}
+#endif
+		} break;
+		case NOTIFICATION_INTERNAL_PROCESS: {
+			ERR_FAIL_COND(!enabled);
```
Comment (fire) [inline scene/2d/skeleton_modification_2d.cpp:152]: See also 2d, but this causes error spam.

**Comment by lyuma** [comment]
I have decided to mark this as draft so that we can focus on other more pressing priorities for 4.0 beta, and fix some of the design issues with this that came up during testing.

For now, I will submit a new PR that marks all of the existing SkeletonModification resources and SkeletonModificationStack as Deprecated (planned to move to modification nodes) and unexposes some skeleton APIs only being used by the modifications such as "global_pose_z_forward_to_bone_forward" so it will be easier to remove

**Comment by fire** [comment]
We're also marking SkeletonIK to Deprecated too?

**Comment by 9gay** [comment]
please add this issue too https://github.com/godotengine/godot/issues/54891

**Review by RedMser — COMMENTED** [review]
I did some initial testing (after fixing merge conflicts) and the new system is a lot more intuitive to use and set up. Great work!

I only tested FABRIK in 3D so far, but it seems like the math is no longer correct, as the target node is not being reached anymore: 
22-10-18_00-30-50_godot windows editor dev x86_64_Warbler

It looks like the algorithm results in global transform per bone, but is applied in local space (see how each bone rotates by 90°): 
image

Let me know if you need a simple repro project.

**Comment by lufog** [comment]
I guess this will close #65265 unless it's an inspector-specific bug.

**Comment by lufog** [comment]
I guess this will close #65265 unless it's an inspector-specific bug.

**Comment by lyuma** [comment]
I have repurposed this PR into #71137 and removed SkeletonModificationStack3D, so we will not port to nodes in time for 4.0.

2D modification stack works and will not be removed for now.

3D modification stack will hopefully return in a future 4.x in a different form. Stay tuned.

**Pull request closed without merging**